{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jackyjacky607/AsiaGodTone-Hotpot-Game/blob/master/SoVITS.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### ➤ [GPT-SoVITS 手把手教學文章](https://dean-lin.medium.com/f6a620cf7fc6)\n",
        "#### ➤ [筆者的 Medium 部落格](https://medium.com/@dean-lin)"
      ],
      "metadata": {
        "id": "sMm5S3CNp2oD"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UtOtTFiGZJ-N"
      },
      "source": [
        "## 0. 初始化檢查"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "maUJ4mNpYxxQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "270e1da8-56d9-4110-d3fb-6a8f2ce29d83"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Default GPU Device Details: /device:GPU:0\n"
          ]
        }
      ],
      "source": [
        "import tensorflow as tf\n",
        "\n",
        "if tf.test.gpu_device_name():\n",
        "    print('Default GPU Device Details: {}'.format(tf.test.gpu_device_name()))\n",
        "else:\n",
        "    raise SystemError('GPU device not found')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hIgovmIrZVAc"
      },
      "source": [
        " ## 1. 掛載到 Google Drive"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "rMytosXDZUMg",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "160da66b-7770-4eac-b0ca-b18eaf6d56c3"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "66ri0484Yz64"
      },
      "source": [
        "## 2. Clone Github 最新程式，並下載對應的環境"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "8rGL8TjVsjUD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4bde615b-a1c5-496f-80b1-6ee4d7279c28"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "fatal: destination path 'GPT-SoVITS' already exists and is not an empty directory.\n",
            "/content/GPT-SoVITS\n",
            "Cloning into 'GPT_SoVITS/pretrained_models'...\n",
            "remote: Enumerating objects: 20, done.\u001b[K\n",
            "remote: Counting objects: 100% (16/16), done.\u001b[K\n",
            "remote: Compressing objects: 100% (16/16), done.\u001b[K\n",
            "remote: Total 20 (delta 1), reused 0 (delta 0), pack-reused 4\u001b[K\n",
            "Unpacking objects: 100% (20/20), 102.34 KiB | 2.84 MiB/s, done.\n",
            "Filtering content: 100% (5/5), 1.11 GiB | 56.19 MiB/s, done.\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 1)) (1.23.5)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 2)) (1.11.4)\n",
            "Requirement already satisfied: tensorboard in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 3)) (2.15.2)\n",
            "Requirement already satisfied: librosa==0.9.2 in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 4)) (0.9.2)\n",
            "Requirement already satisfied: numba==0.56.4 in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 5)) (0.56.4)\n",
            "Requirement already satisfied: pytorch-lightning in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 6)) (2.2.1)\n",
            "Collecting gradio==3.38.0 (from -r requirements.txt (line 7))\n",
            "  Using cached gradio-3.38.0-py3-none-any.whl (19.8 MB)\n",
            "Collecting gradio_client==0.8.1 (from -r requirements.txt (line 8))\n",
            "  Using cached gradio_client-0.8.1-py3-none-any.whl (305 kB)\n",
            "Requirement already satisfied: ffmpeg-python in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 9)) (0.2.0)\n",
            "Requirement already satisfied: onnxruntime in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 10)) (1.17.1)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 11)) (4.66.2)\n",
            "Requirement already satisfied: funasr==1.0.0 in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 12)) (1.0.0)\n",
            "Requirement already satisfied: cn2an in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 13)) (0.5.22)\n",
            "Requirement already satisfied: pypinyin in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 14)) (0.51.0)\n",
            "Requirement already satisfied: pyopenjtalk in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 15)) (0.3.3)\n",
            "Requirement already satisfied: g2p_en in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 16)) (2.1.0)\n",
            "Requirement already satisfied: torchaudio in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 17)) (2.2.1+cu121)\n",
            "Requirement already satisfied: modelscope==1.10.0 in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 18)) (1.10.0)\n",
            "Requirement already satisfied: sentencepiece in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 19)) (0.1.99)\n",
            "Requirement already satisfied: transformers in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 20)) (4.38.2)\n",
            "Requirement already satisfied: chardet in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 21)) (5.2.0)\n",
            "Requirement already satisfied: PyYAML in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 22)) (6.0.1)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 23)) (5.9.5)\n",
            "Requirement already satisfied: jieba_fast in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 24)) (0.53)\n",
            "Requirement already satisfied: jieba in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 25)) (0.42.1)\n",
            "Requirement already satisfied: LangSegment in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 26)) (0.3.0)\n",
            "Requirement already satisfied: Faster_Whisper in /usr/local/lib/python3.10/dist-packages (from -r requirements.txt (line 27)) (1.0.1)\n",
            "Requirement already satisfied: audioread>=2.1.9 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (3.0.1)\n",
            "Requirement already satisfied: scikit-learn>=0.19.1 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (1.2.2)\n",
            "Requirement already satisfied: joblib>=0.14 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (1.3.2)\n",
            "Requirement already satisfied: decorator>=4.0.10 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (4.4.2)\n",
            "Requirement already satisfied: resampy>=0.2.2 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (0.4.3)\n",
            "Requirement already satisfied: soundfile>=0.10.2 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (0.12.1)\n",
            "Requirement already satisfied: pooch>=1.0 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (1.8.1)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from librosa==0.9.2->-r requirements.txt (line 4)) (24.0)\n",
            "Requirement already satisfied: llvmlite<0.40,>=0.39.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba==0.56.4->-r requirements.txt (line 5)) (0.39.1)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from numba==0.56.4->-r requirements.txt (line 5)) (67.7.2)\n",
            "Requirement already satisfied: aiofiles<24.0,>=22.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (23.2.1)\n",
            "Requirement already satisfied: aiohttp~=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (3.9.3)\n",
            "Requirement already satisfied: altair<6.0,>=4.2.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (4.2.2)\n",
            "Requirement already satisfied: fastapi in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.110.1)\n",
            "Requirement already satisfied: ffmpy in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.3.2)\n",
            "Requirement already satisfied: httpx in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.27.0)\n",
            "Requirement already satisfied: huggingface-hub>=0.14.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.20.3)\n",
            "Requirement already satisfied: jinja2<4.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (3.1.3)\n",
            "Requirement already satisfied: markdown-it-py[linkify]>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.2.0)\n",
            "Requirement already satisfied: markupsafe~=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.1.5)\n",
            "Requirement already satisfied: matplotlib~=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (3.7.1)\n",
            "Requirement already satisfied: mdit-py-plugins<=0.3.3 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.3.3)\n",
            "Requirement already satisfied: orjson~=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (3.10.0)\n",
            "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.0.3)\n",
            "Requirement already satisfied: pillow<11.0,>=8.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (9.4.0)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,<3.0.0,>=1.7.4 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.6.4)\n",
            "Requirement already satisfied: pydub in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.25.1)\n",
            "Requirement already satisfied: python-multipart in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.0.9)\n",
            "Requirement already satisfied: requests~=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.31.0)\n",
            "Requirement already satisfied: semantic-version~=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (2.10.0)\n",
            "Requirement already satisfied: typing-extensions~=4.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (4.10.0)\n",
            "Requirement already satisfied: uvicorn>=0.14.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (0.29.0)\n",
            "Requirement already satisfied: websockets<12.0,>=10.0 in /usr/local/lib/python3.10/dist-packages (from gradio==3.38.0->-r requirements.txt (line 7)) (11.0.3)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from gradio_client==0.8.1->-r requirements.txt (line 8)) (2023.6.0)\n",
            "Requirement already satisfied: jamo in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.4.1)\n",
            "Requirement already satisfied: kaldiio>=2.17.0 in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (2.18.0)\n",
            "Requirement already satisfied: torch-complex in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.4.3)\n",
            "Requirement already satisfied: pytorch-wpe in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.0.1)\n",
            "Requirement already satisfied: editdistance>=0.5.2 in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.6.2)\n",
            "Requirement already satisfied: oss2 in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (2.18.4)\n",
            "Requirement already satisfied: hdbscan in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.8.33)\n",
            "Requirement already satisfied: umap in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.1.1)\n",
            "Requirement already satisfied: jaconv in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (0.3.4)\n",
            "Requirement already satisfied: hydra-core>=1.3.2 in /usr/local/lib/python3.10/dist-packages (from funasr==1.0.0->-r requirements.txt (line 12)) (1.3.2)\n",
            "Requirement already satisfied: addict in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (2.4.0)\n",
            "Requirement already satisfied: attrs in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (23.2.0)\n",
            "Requirement already satisfied: datasets>=2.14.5 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (2.18.0)\n",
            "Requirement already satisfied: einops in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (0.7.0)\n",
            "Requirement already satisfied: filelock>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (3.13.3)\n",
            "Requirement already satisfied: gast>=0.2.2 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (0.5.4)\n",
            "Requirement already satisfied: pyarrow!=9.0.0,>=6.0.0 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (14.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (2.8.2)\n",
            "Requirement already satisfied: simplejson>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (3.19.2)\n",
            "Requirement already satisfied: sortedcontainers>=1.5.9 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (2.4.0)\n",
            "Requirement already satisfied: urllib3>=1.26 in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (2.0.7)\n",
            "Requirement already satisfied: yapf in /usr/local/lib/python3.10/dist-packages (from modelscope==1.10.0->-r requirements.txt (line 18)) (0.40.2)\n",
            "Requirement already satisfied: absl-py>=0.4 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (1.4.0)\n",
            "Requirement already satisfied: grpcio>=1.48.2 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (1.62.1)\n",
            "Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (2.27.0)\n",
            "Requirement already satisfied: google-auth-oauthlib<2,>=0.5 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (1.2.0)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (3.6)\n",
            "Requirement already satisfied: protobuf!=4.24.0,>=3.19.6 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (3.20.3)\n",
            "Requirement already satisfied: six>1.9 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (1.16.0)\n",
            "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (0.7.2)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard->-r requirements.txt (line 3)) (3.0.2)\n",
            "Requirement already satisfied: torch>=1.13.0 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning->-r requirements.txt (line 6)) (2.2.1+cu121)\n",
            "Requirement already satisfied: torchmetrics>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning->-r requirements.txt (line 6)) (1.3.2)\n",
            "Requirement already satisfied: lightning-utilities>=0.8.0 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning->-r requirements.txt (line 6)) (0.11.2)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.10/dist-packages (from ffmpeg-python->-r requirements.txt (line 9)) (0.18.3)\n",
            "Requirement already satisfied: coloredlogs in /usr/local/lib/python3.10/dist-packages (from onnxruntime->-r requirements.txt (line 10)) (15.0.1)\n",
            "Requirement already satisfied: flatbuffers in /usr/local/lib/python3.10/dist-packages (from onnxruntime->-r requirements.txt (line 10)) (24.3.25)\n",
            "Requirement already satisfied: sympy in /usr/local/lib/python3.10/dist-packages (from onnxruntime->-r requirements.txt (line 10)) (1.12)\n",
            "Requirement already satisfied: proces>=0.1.3 in /usr/local/lib/python3.10/dist-packages (from cn2an->-r requirements.txt (line 13)) (0.1.7)\n",
            "Requirement already satisfied: nltk>=3.2.4 in /usr/local/lib/python3.10/dist-packages (from g2p_en->-r requirements.txt (line 16)) (3.8.1)\n",
            "Requirement already satisfied: inflect>=0.3.1 in /usr/local/lib/python3.10/dist-packages (from g2p_en->-r requirements.txt (line 16)) (7.0.0)\n",
            "Requirement already satisfied: distance>=0.1.3 in /usr/local/lib/python3.10/dist-packages (from g2p_en->-r requirements.txt (line 16)) (0.1.3)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (3.2.1)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.105)\n",
            "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (8.9.2.26)\n",
            "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.3.1)\n",
            "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (11.0.2.54)\n",
            "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (10.3.2.106)\n",
            "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (11.4.5.107)\n",
            "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.0.106)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.19.3 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (2.19.3)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.1.105)\n",
            "Requirement already satisfied: triton==2.2.0 in /usr/local/lib/python3.10/dist-packages (from torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (2.2.0)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12 in /usr/local/lib/python3.10/dist-packages (from nvidia-cusolver-cu12==11.4.5.107->torch>=1.13.0->pytorch-lightning->-r requirements.txt (line 6)) (12.4.127)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.10/dist-packages (from transformers->-r requirements.txt (line 20)) (2023.12.25)\n",
            "Requirement already satisfied: tokenizers<0.19,>=0.14 in /usr/local/lib/python3.10/dist-packages (from transformers->-r requirements.txt (line 20)) (0.15.2)\n",
            "Requirement already satisfied: safetensors>=0.4.1 in /usr/local/lib/python3.10/dist-packages (from transformers->-r requirements.txt (line 20)) (0.4.2)\n",
            "Requirement already satisfied: py3langid>=0.2.2 in /usr/local/lib/python3.10/dist-packages (from LangSegment->-r requirements.txt (line 26)) (0.2.2)\n",
            "Requirement already satisfied: av==11.* in /usr/local/lib/python3.10/dist-packages (from Faster_Whisper->-r requirements.txt (line 27)) (11.0.0)\n",
            "Requirement already satisfied: ctranslate2<5,>=4.0 in /usr/local/lib/python3.10/dist-packages (from Faster_Whisper->-r requirements.txt (line 27)) (4.1.0)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.3.1)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.4.1)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (6.0.5)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.9.4)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (4.0.3)\n",
            "Requirement already satisfied: entrypoints in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.4)\n",
            "Requirement already satisfied: jsonschema>=3.0 in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (4.19.2)\n",
            "Requirement already satisfied: toolz in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.12.1)\n",
            "Requirement already satisfied: pyarrow-hotfix in /usr/local/lib/python3.10/dist-packages (from datasets>=2.14.5->modelscope==1.10.0->-r requirements.txt (line 18)) (0.6)\n",
            "Requirement already satisfied: dill<0.3.9,>=0.3.0 in /usr/local/lib/python3.10/dist-packages (from datasets>=2.14.5->modelscope==1.10.0->-r requirements.txt (line 18)) (0.3.8)\n",
            "Requirement already satisfied: xxhash in /usr/local/lib/python3.10/dist-packages (from datasets>=2.14.5->modelscope==1.10.0->-r requirements.txt (line 18)) (3.4.1)\n",
            "Requirement already satisfied: multiprocess in /usr/local/lib/python3.10/dist-packages (from datasets>=2.14.5->modelscope==1.10.0->-r requirements.txt (line 18)) (0.70.16)\n",
            "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard->-r requirements.txt (line 3)) (5.3.3)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard->-r requirements.txt (line 3)) (0.4.0)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.10/dist-packages (from google-auth<3,>=1.6.3->tensorboard->-r requirements.txt (line 3)) (4.9)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from google-auth-oauthlib<2,>=0.5->tensorboard->-r requirements.txt (line 3)) (1.3.1)\n",
            "Requirement already satisfied: omegaconf<2.4,>=2.2 in /usr/local/lib/python3.10/dist-packages (from hydra-core>=1.3.2->funasr==1.0.0->-r requirements.txt (line 12)) (2.3.0)\n",
            "Requirement already satisfied: antlr4-python3-runtime==4.9.* in /usr/local/lib/python3.10/dist-packages (from hydra-core>=1.3.2->funasr==1.0.0->-r requirements.txt (line 12)) (4.9.3)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py[linkify]>=2.0.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.1.2)\n",
            "Requirement already satisfied: linkify-it-py<3,>=1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py[linkify]>=2.0.0->gradio==3.38.0->-r requirements.txt (line 7)) (2.0.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (4.50.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.4.5)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio==3.38.0->-r requirements.txt (line 7)) (3.1.2)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from nltk>=3.2.4->g2p_en->-r requirements.txt (line 16)) (8.1.7)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio==3.38.0->-r requirements.txt (line 7)) (2023.4)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio==3.38.0->-r requirements.txt (line 7)) (2024.1)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch>=1.0->librosa==0.9.2->-r requirements.txt (line 4)) (4.2.0)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,<3.0.0,>=1.7.4->gradio==3.38.0->-r requirements.txt (line 7)) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.16.3 in /usr/local/lib/python3.10/dist-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,<3.0.0,>=1.7.4->gradio==3.38.0->-r requirements.txt (line 7)) (2.16.3)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests~=2.0->gradio==3.38.0->-r requirements.txt (line 7)) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests~=2.0->gradio==3.38.0->-r requirements.txt (line 7)) (3.6)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests~=2.0->gradio==3.38.0->-r requirements.txt (line 7)) (2024.2.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=0.19.1->librosa==0.9.2->-r requirements.txt (line 4)) (3.4.0)\n",
            "Requirement already satisfied: cffi>=1.0 in /usr/local/lib/python3.10/dist-packages (from soundfile>=0.10.2->librosa==0.9.2->-r requirements.txt (line 4)) (1.16.0)\n",
            "Requirement already satisfied: h11>=0.8 in /usr/local/lib/python3.10/dist-packages (from uvicorn>=0.14.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.14.0)\n",
            "Requirement already satisfied: humanfriendly>=9.1 in /usr/local/lib/python3.10/dist-packages (from coloredlogs->onnxruntime->-r requirements.txt (line 10)) (10.0)\n",
            "Requirement already satisfied: starlette<0.38.0,>=0.37.2 in /usr/local/lib/python3.10/dist-packages (from fastapi->gradio==3.38.0->-r requirements.txt (line 7)) (0.37.2)\n",
            "Requirement already satisfied: cython<3,>=0.27 in /usr/local/lib/python3.10/dist-packages (from hdbscan->funasr==1.0.0->-r requirements.txt (line 12)) (0.29.37)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.10/dist-packages (from httpx->gradio==3.38.0->-r requirements.txt (line 7)) (3.7.1)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.10/dist-packages (from httpx->gradio==3.38.0->-r requirements.txt (line 7)) (1.0.5)\n",
            "Requirement already satisfied: sniffio in /usr/local/lib/python3.10/dist-packages (from httpx->gradio==3.38.0->-r requirements.txt (line 7)) (1.3.1)\n",
            "Requirement already satisfied: crcmod>=1.7 in /usr/local/lib/python3.10/dist-packages (from oss2->funasr==1.0.0->-r requirements.txt (line 12)) (1.7)\n",
            "Requirement already satisfied: pycryptodome>=3.4.7 in /usr/local/lib/python3.10/dist-packages (from oss2->funasr==1.0.0->-r requirements.txt (line 12)) (3.20.0)\n",
            "Requirement already satisfied: aliyun-python-sdk-kms>=2.4.1 in /usr/local/lib/python3.10/dist-packages (from oss2->funasr==1.0.0->-r requirements.txt (line 12)) (2.16.2)\n",
            "Requirement already satisfied: aliyun-python-sdk-core>=2.13.12 in /usr/local/lib/python3.10/dist-packages (from oss2->funasr==1.0.0->-r requirements.txt (line 12)) (2.15.0)\n",
            "Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.10/dist-packages (from sympy->onnxruntime->-r requirements.txt (line 10)) (1.3.0)\n",
            "Requirement already satisfied: importlib-metadata>=6.6.0 in /usr/local/lib/python3.10/dist-packages (from yapf->modelscope==1.10.0->-r requirements.txt (line 18)) (7.1.0)\n",
            "Requirement already satisfied: tomli>=2.0.1 in /usr/local/lib/python3.10/dist-packages (from yapf->modelscope==1.10.0->-r requirements.txt (line 18)) (2.0.1)\n",
            "Requirement already satisfied: jmespath<1.0.0,>=0.9.3 in /usr/local/lib/python3.10/dist-packages (from aliyun-python-sdk-core>=2.13.12->oss2->funasr==1.0.0->-r requirements.txt (line 12)) (0.10.0)\n",
            "Requirement already satisfied: cryptography>=2.6.0 in /usr/local/lib/python3.10/dist-packages (from aliyun-python-sdk-core>=2.13.12->oss2->funasr==1.0.0->-r requirements.txt (line 12)) (42.0.5)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.10/dist-packages (from cffi>=1.0->soundfile>=0.10.2->librosa==0.9.2->-r requirements.txt (line 4)) (2.22)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=6.6.0->yapf->modelscope==1.10.0->-r requirements.txt (line 18)) (3.18.1)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (2023.12.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.34.0)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio==3.38.0->-r requirements.txt (line 7)) (0.18.0)\n",
            "Requirement already satisfied: uc-micro-py in /usr/local/lib/python3.10/dist-packages (from linkify-it-py<3,>=1->markdown-it-py[linkify]>=2.0.0->gradio==3.38.0->-r requirements.txt (line 7)) (1.0.3)\n",
            "Requirement already satisfied: pyasn1<0.7.0,>=0.4.6 in /usr/local/lib/python3.10/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard->-r requirements.txt (line 3)) (0.6.0)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<2,>=0.5->tensorboard->-r requirements.txt (line 3)) (3.2.2)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio->httpx->gradio==3.38.0->-r requirements.txt (line 7)) (1.2.0)\n",
            "Installing collected packages: gradio_client, gradio\n",
            "  Attempting uninstall: gradio_client\n",
            "    Found existing installation: gradio_client 0.15.0\n",
            "    Uninstalling gradio_client-0.15.0:\n",
            "      Successfully uninstalled gradio_client-0.15.0\n",
            "  Attempting uninstall: gradio\n",
            "    Found existing installation: gradio 4.25.0\n",
            "    Uninstalling gradio-4.25.0:\n",
            "      Successfully uninstalled gradio-4.25.0\n",
            "Successfully installed gradio-3.38.0 gradio_client-0.8.1\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "ffmpeg is already the newest version (7:4.4.2-0ubuntu0.22.04.1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 45 not upgraded.\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "libsox-dev is already the newest version (14.4.2+git20190427-2+deb11u2ubuntu0.22.04.1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 45 not upgraded.\n",
            "Requirement already satisfied: ffmpeg-python==0.2.0 in /usr/local/lib/python3.10/dist-packages (0.2.0)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.10/dist-packages (from ffmpeg-python==0.2.0) (0.18.3)\n",
            "Requirement already satisfied: gradio in /usr/local/lib/python3.10/dist-packages (3.38.0)\n",
            "Collecting gradio\n",
            "  Using cached gradio-4.25.0-py3-none-any.whl (17.1 MB)\n",
            "Requirement already satisfied: aiofiles<24.0,>=22.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (23.2.1)\n",
            "Requirement already satisfied: altair<6.0,>=4.2.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (4.2.2)\n",
            "Requirement already satisfied: fastapi in /usr/local/lib/python3.10/dist-packages (from gradio) (0.110.1)\n",
            "Requirement already satisfied: ffmpy in /usr/local/lib/python3.10/dist-packages (from gradio) (0.3.2)\n",
            "Collecting gradio-client==0.15.0 (from gradio)\n",
            "  Using cached gradio_client-0.15.0-py3-none-any.whl (313 kB)\n",
            "Requirement already satisfied: httpx>=0.24.1 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.27.0)\n",
            "Requirement already satisfied: huggingface-hub>=0.19.3 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.20.3)\n",
            "Requirement already satisfied: importlib-resources<7.0,>=1.3 in /usr/local/lib/python3.10/dist-packages (from gradio) (6.4.0)\n",
            "Requirement already satisfied: jinja2<4.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (3.1.3)\n",
            "Requirement already satisfied: markupsafe~=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.1.5)\n",
            "Requirement already satisfied: matplotlib~=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (3.7.1)\n",
            "Requirement already satisfied: numpy~=1.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (1.23.5)\n",
            "Requirement already satisfied: orjson~=3.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (3.10.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from gradio) (24.0)\n",
            "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.0.3)\n",
            "Requirement already satisfied: pillow<11.0,>=8.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (9.4.0)\n",
            "Requirement already satisfied: pydantic>=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.6.4)\n",
            "Requirement already satisfied: pydub in /usr/local/lib/python3.10/dist-packages (from gradio) (0.25.1)\n",
            "Requirement already satisfied: python-multipart>=0.0.9 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.0.9)\n",
            "Requirement already satisfied: pyyaml<7.0,>=5.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (6.0.1)\n",
            "Requirement already satisfied: ruff>=0.2.2 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.3.5)\n",
            "Requirement already satisfied: semantic-version~=2.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (2.10.0)\n",
            "Requirement already satisfied: tomlkit==0.12.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.12.0)\n",
            "Requirement already satisfied: typer[all]<1.0,>=0.9 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.9.4)\n",
            "Requirement already satisfied: typing-extensions~=4.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (4.10.0)\n",
            "Requirement already satisfied: uvicorn>=0.14.0 in /usr/local/lib/python3.10/dist-packages (from gradio) (0.29.0)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from gradio-client==0.15.0->gradio) (2023.6.0)\n",
            "Requirement already satisfied: websockets<12.0,>=10.0 in /usr/local/lib/python3.10/dist-packages (from gradio-client==0.15.0->gradio) (11.0.3)\n",
            "Requirement already satisfied: entrypoints in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio) (0.4)\n",
            "Requirement already satisfied: jsonschema>=3.0 in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio) (4.19.2)\n",
            "Requirement already satisfied: toolz in /usr/local/lib/python3.10/dist-packages (from altair<6.0,>=4.2.0->gradio) (0.12.1)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (3.7.1)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (2024.2.2)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (1.0.5)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (3.6)\n",
            "Requirement already satisfied: sniffio in /usr/local/lib/python3.10/dist-packages (from httpx>=0.24.1->gradio) (1.3.1)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.10/dist-packages (from httpcore==1.*->httpx>=0.24.1->gradio) (0.14.0)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.19.3->gradio) (3.13.3)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.19.3->gradio) (2.31.0)\n",
            "Requirement already satisfied: tqdm>=4.42.1 in /usr/local/lib/python3.10/dist-packages (from huggingface-hub>=0.19.3->gradio) (4.66.2)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (1.2.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (4.50.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (1.4.5)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (3.1.2)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib~=3.0->gradio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio) (2023.4)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas<3.0,>=1.0->gradio) (2024.1)\n",
            "Requirement already satisfied: annotated-types>=0.4.0 in /usr/local/lib/python3.10/dist-packages (from pydantic>=2.0->gradio) (0.6.0)\n",
            "Requirement already satisfied: pydantic-core==2.16.3 in /usr/local/lib/python3.10/dist-packages (from pydantic>=2.0->gradio) (2.16.3)\n",
            "Requirement already satisfied: click<9.0.0,>=7.1.1 in /usr/local/lib/python3.10/dist-packages (from typer[all]<1.0,>=0.9->gradio) (8.1.7)\n",
            "Requirement already satisfied: colorama<0.5.0,>=0.4.3 in /usr/local/lib/python3.10/dist-packages (from typer[all]<1.0,>=0.9->gradio) (0.4.6)\n",
            "Requirement already satisfied: shellingham<2.0.0,>=1.3.0 in /usr/local/lib/python3.10/dist-packages (from typer[all]<1.0,>=0.9->gradio) (1.5.4)\n",
            "Requirement already satisfied: rich<14.0.0,>=10.11.0 in /usr/local/lib/python3.10/dist-packages (from typer[all]<1.0,>=0.9->gradio) (13.7.1)\n",
            "Requirement already satisfied: starlette<0.38.0,>=0.37.2 in /usr/local/lib/python3.10/dist-packages (from fastapi->gradio) (0.37.2)\n",
            "Requirement already satisfied: attrs>=22.2.0 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio) (23.2.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio) (2023.12.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio) (0.34.0)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.10/dist-packages (from jsonschema>=3.0->altair<6.0,>=4.2.0->gradio) (0.18.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib~=3.0->gradio) (1.16.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich<14.0.0,>=10.11.0->typer[all]<1.0,>=0.9->gradio) (2.2.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich<14.0.0,>=10.11.0->typer[all]<1.0,>=0.9->gradio) (2.16.1)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anyio->httpx>=0.24.1->gradio) (1.2.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.19.3->gradio) (3.3.2)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->huggingface-hub>=0.19.3->gradio) (2.0.7)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich<14.0.0,>=10.11.0->typer[all]<1.0,>=0.9->gradio) (0.1.2)\n",
            "Installing collected packages: gradio-client, gradio\n",
            "  Attempting uninstall: gradio-client\n",
            "    Found existing installation: gradio_client 0.8.1\n",
            "    Uninstalling gradio_client-0.8.1:\n",
            "      Successfully uninstalled gradio_client-0.8.1\n",
            "  Attempting uninstall: gradio\n",
            "    Found existing installation: gradio 3.38.0\n",
            "    Uninstalling gradio-3.38.0:\n",
            "      Successfully uninstalled gradio-3.38.0\n",
            "Successfully installed gradio-4.25.0 gradio-client-0.15.0\n"
          ]
        }
      ],
      "source": [
        "!git clone https://github.com/dean9703111/GPT-SoVITS # 我改成自己 Fork 出來舊版的，官方新版不太穩定\n",
        "!rm -rf GPT_SoVITS/pretrained_models  # 刪除目標資料夾\n",
        "!git clone https://huggingface.co/lj1995/GPT-SoVITS GPT_SoVITS/pretrained_models  # 複製模型\n",
        "!pip install -r requirements.txt\n",
        "!sudo apt install ffmpeg\n",
        "!sudo apt install libsox-dev\n",
        "!pip install ffmpeg-python==0.2.0\n",
        "!pip install --upgrade gradio"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9ZRVt9Kbcrma"
      },
      "source": [
        "## 3. 啟動程式\n",
        "點擊 Running on public URL: xxx，即可進入 Web UI 畫面"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import re\n",
        "file_path = '/content/GPT-SoVITS/webui.py'\n",
        "with open(file_path, 'r') as file:\n",
        "    content = file.read()\n",
        "modified_content = content.replace(\"app.queue(concurrency_count=511, max_size=1022)\", \"app.queue(max_size=1022)\")\n",
        "with open(file_path, 'w') as file:\n",
        "    file.write(modified_content)\n",
        "file_path = '/content/GPT-SoVITS/GPT_SoVITS/inference_webui.py'\n",
        "with open(file_path, 'r') as file:\n",
        "    content = file.read()\n",
        "modified_content = content.replace(\"app.queue(concurrency_count=511, max_size=1022)\", \"app.queue(max_size=1022)\")\n",
        "with open(file_path, 'w') as file:\n",
        "    file.write(modified_content)"
      ],
      "metadata": {
        "id": "rTUzP9qwJF7Q"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!sed -i 's/is_share=False/is_share=True/g' config.py\n",
        "!python webui.py"
      ],
      "metadata": {
        "id": "s5xoT-62aXe8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bb3cc07b-2c83-46f7-94c0-38f57623468b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running on public URL: https://74f9ef7c75d1a27c67.gradio.live\n",
            "\"/usr/bin/python3\" tools/slice_audio.py \"/content/drive/MyDrive/voices/raw\" \"/content/drive/MyDrive/voices/slicer\" -34 4000 300 10 500 0.9 0.25 0 4\n",
            "\"/usr/bin/python3\" tools/slice_audio.py \"/content/drive/MyDrive/voices/raw\" \"/content/drive/MyDrive/voices/slicer\" -34 4000 300 10 500 0.9 0.25 1 4\n",
            "\"/usr/bin/python3\" tools/slice_audio.py \"/content/drive/MyDrive/voices/raw\" \"/content/drive/MyDrive/voices/slicer\" -34 4000 300 10 500 0.9 0.25 2 4\n",
            "\"/usr/bin/python3\" tools/slice_audio.py \"/content/drive/MyDrive/voices/raw\" \"/content/drive/MyDrive/voices/slicer\" -34 4000 300 10 500 0.9 0.25 3 4\n",
            "执行完毕，请检查输出文件\n",
            "执行完毕，请检查输出文件\n",
            "执行完毕，请检查输出文件\n",
            "执行完毕，请检查输出文件\n",
            "\"/usr/bin/python3\" tools/asr/funasr_asr.py -i \"/content/drive/MyDrive/voices/slicer\" -o \"output/asr_opt\" -s large -l zh -p float16\n",
            "Please install rotary_embedding_torch by: \n",
            " pip install -U rotary_embedding_torch\n",
            "Please install rotary_embedding_torch by: \n",
            " pip install -U rotary_embedding_torch\n",
            "Please install rotary_embedding_torch by: \n",
            " pip install -U rotary_embedding_torch\n",
            "Please install rotary_embedding_torch by: \n",
            " pip install -U rotary_embedding_torch\n",
            "\n",
            "tables: \n",
            "\n",
            "-----------    ** dataset_classes **    --------------\n",
            "| class name   | class location                               |\n",
            "| AudioDataset | funasr/datasets/audio_datasets/datasets.py:7 |\n",
            "-----------    ** index_ds_classes **    --------------\n",
            "| class name   | class location                               |\n",
            "| IndexDSJsonl | funasr/datasets/audio_datasets/index_ds.py:9 |\n",
            "-----------    ** batch_sampler_classes **    --------------\n",
            "| class name   | class location                               |\n",
            "| BatchSampler | funasr/datasets/audio_datasets/samplers.py:7 |\n",
            "-----------    ** frontend_classes **    --------------\n",
            "| class name        | class location                       |\n",
            "| WavFrontend       | funasr/frontends/wav_frontend.py:78  |\n",
            "| WavFrontendOnline | funasr/frontends/wav_frontend.py:216 |\n",
            "-----------    ** encoder_classes **    --------------\n",
            "| class name            | class location                                        |\n",
            "| BranchformerEncoder   | funasr/models/branchformer/encoder.py:294             |\n",
            "| ConformerChunkEncoder | funasr/models/bat/conformer_chunk_encoder.py:315      |\n",
            "| ConformerEncoder      | funasr/models/conformer/encoder.py:286                |\n",
            "| DFSMN                 | funasr/models/fsmn_vad_streaming/encoder.py:232       |\n",
            "| EBranchformerEncoder  | funasr/models/e_branchformer/encoder.py:177           |\n",
            "| FSMN                  | funasr/models/fsmn_vad_streaming/encoder.py:161       |\n",
            "| SANMEncoder           | funasr/models/sanm/encoder.py:161                     |\n",
            "| SANMEncoderChunkOpt   | funasr/models/scama/encoder.py:162                    |\n",
            "| SANMVadEncoder        | funasr/models/ct_transformer_streaming/encoder.py:148 |\n",
            "| TransformerEncoder    | funasr/models/transformer/encoder.py:139              |\n",
            "-----------    ** predictor_classes **    --------------\n",
            "| class name     | class location                                     |\n",
            "| CifPredictor   | funasr/models/paraformer/cif_predictor.py:15       |\n",
            "| CifPredictorV2 | funasr/models/paraformer/cif_predictor.py:141      |\n",
            "| CifPredictorV3 | funasr/models/bicif_paraformer/cif_predictor.py:95 |\n",
            "-----------    ** model_classes **    --------------\n",
            "| class name             | class location                                     |\n",
            "| BiCifParaformer        | funasr/models/bicif_paraformer/model.py:37         |\n",
            "| Branchformer           | funasr/models/branchformer/model.py:6              |\n",
            "| CAMPPlus               | funasr/models/campplus/model.py:30                 |\n",
            "| CTTransformer          | funasr/models/ct_transformer/model.py:30           |\n",
            "| CTTransformerStreaming | funasr/models/ct_transformer_streaming/model.py:27 |\n",
            "| Conformer              | funasr/models/conformer/model.py:8                 |\n",
            "| ContextualParaformer   | funasr/models/contextual_paraformer/model.py:43    |\n",
            "| EBranchformer          | funasr/models/e_branchformer/model.py:6            |\n",
            "| Emotion2vec            | funasr/models/emotion2vec/model.py:34              |\n",
            "| FsmnVADStreaming       | funasr/models/fsmn_vad_streaming/model.py:267      |\n",
            "| MonotonicAligner       | funasr/models/monotonic_aligner/model.py:24        |\n",
            "| Paraformer             | funasr/models/paraformer/model.py:26               |\n",
            "| ParaformerStreaming    | funasr/models/paraformer_streaming/model.py:37     |\n",
            "| SANM                   | funasr/models/sanm/model.py:13                     |\n",
            "| SCAMA                  | funasr/models/scama/model.py:38                    |\n",
            "| SeacoParaformer        | funasr/models/seaco_paraformer/model.py:45         |\n",
            "| Transformer            | funasr/models/transformer/model.py:20              |\n",
            "| UniASR                 | funasr/models/uniasr/model.py:26                   |\n",
            "-----------    ** decoder_classes **    --------------\n",
            "| class name                                 | class location                                     |\n",
            "| ContextualParaformerDecoder                | funasr/models/contextual_paraformer/decoder.py:103 |\n",
            "| DynamicConvolution2DTransformerDecoder     | funasr/models/transformer/decoder.py:588           |\n",
            "| DynamicConvolutionTransformerDecoder       | funasr/models/transformer/decoder.py:527           |\n",
            "| FsmnDecoder                                | funasr/models/sanm/decoder.py:198                  |\n",
            "| FsmnDecoderSCAMAOpt                        | funasr/models/scama/decoder.py:197                 |\n",
            "| LightweightConvolution2DTransformerDecoder | funasr/models/transformer/decoder.py:465           |\n",
            "| LightweightConvolutionTransformerDecoder   | funasr/models/transformer/decoder.py:404           |\n",
            "| ParaformerSANDecoder                       | funasr/models/paraformer/decoder.py:529            |\n",
            "| ParaformerSANMDecoder                      | funasr/models/paraformer/decoder.py:204            |\n",
            "| TransformerDecoder                         | funasr/models/transformer/decoder.py:355           |\n",
            "-----------    ** normalize_classes **    --------------\n",
            "| class name   | class location                             |\n",
            "| GlobalMVN    | funasr/models/normalize/global_mvn.py:11   |\n",
            "| UtteranceMVN | funasr/models/normalize/utterance_mvn.py:8 |\n",
            "-----------    ** specaug_classes **    --------------\n",
            "| class name | class location                       |\n",
            "| SpecAug    | funasr/models/specaug/specaug.py:14  |\n",
            "| SpecAugLFR | funasr/models/specaug/specaug.py:104 |\n",
            "-----------    ** tokenizer_classes **    --------------\n",
            "| class name    | class location                        |\n",
            "| CharTokenizer | funasr/tokenizer/char_tokenizer.py:10 |\n",
            "\n",
            "\n",
            "2024-04-05 05:05:49,128 - modelscope - INFO - PyTorch version 2.2.1+cu121 Found.\n",
            "2024-04-05 05:05:49,130 - modelscope - INFO - TensorFlow version 2.15.0 Found.\n",
            "2024-04-05 05:05:49,130 - modelscope - INFO - Loading ast index from /root/.cache/modelscope/ast_indexer\n",
            "2024-04-05 05:05:49,130 - modelscope - INFO - No valid ast index found from /root/.cache/modelscope/ast_indexer, generating ast index from prebuilt!\n",
            "2024-04-05 05:05:49,222 - modelscope - INFO - Loading done! Current index file version is 1.10.0, with md5 9983905d5646ec47663b4a80242eb94e and a total number of 946 components indexed\n",
            "2024-04-05 05:05:50,216 - modelscope - INFO - Use user-specified model revision: v2.0.4\n",
            "Downloading: 100% 10.9k/10.9k [00:00<00:00, 39.7MB/s]\n",
            "Downloading: 100% 173k/173k [00:00<00:00, 3.61MB/s]\n",
            "Downloading: 100% 2.45k/2.45k [00:00<00:00, 10.9MB/s]\n",
            "Downloading: 100% 472/472 [00:00<00:00, 2.10MB/s]\n",
            "Downloading: 100% 840M/840M [00:05<00:00, 164MB/s]\n",
            "Downloading: 100% 19.1k/19.1k [00:00<00:00, 1.21MB/s]\n",
            "Downloading: 100% 7.90M/7.90M [00:00<00:00, 47.4MB/s]\n",
            "Downloading: 100% 48.7k/48.7k [00:00<00:00, 1.60MB/s]\n",
            "Downloading: 100% 91.5k/91.5k [00:00<00:00, 1.90MB/s]\n",
            "2024-04-05 05:06:11,804 - modelscope - INFO - Use user-specified model revision: v2.0.4\n",
            "Downloading: 100% 7.85k/7.85k [00:00<00:00, 25.0MB/s]\n",
            "Downloading: 100% 1.19k/1.19k [00:00<00:00, 5.63MB/s]\n",
            "Downloading: 100% 365/365 [00:00<00:00, 1.63MB/s]\n",
            "Downloading: 100% 1.64M/1.64M [00:00<00:00, 14.8MB/s]\n",
            "Downloading: 100% 8.45k/8.45k [00:00<00:00, 29.1MB/s]\n",
            "Downloading: 100% 27.3k/27.3k [00:00<00:00, 1.82MB/s]\n",
            "Downloading: 100% 2.16M/2.16M [00:00<00:00, 21.6MB/s]\n",
            "2024-04-05 05:06:20,107 - modelscope - INFO - Use user-specified model revision: v2.0.4\n",
            "Downloading: 100% 6.00k/6.00k [00:00<00:00, 13.3MB/s]\n",
            "Downloading: 100% 810/810 [00:00<00:00, 4.77MB/s]\n",
            "Downloading: 100% 373/373 [00:00<00:00, 2.12MB/s]\n",
            "Downloading: 100% 278M/278M [00:02<00:00, 143MB/s]\n",
            "Downloading: 100% 863/863 [00:00<00:00, 4.27MB/s]\n",
            "Downloading: 100% 11.2k/11.2k [00:00<00:00, 20.3MB/s]\n",
            "Downloading: 100% 151k/151k [00:00<00:00, 3.17MB/s]\n",
            "Downloading: 100% 4.01M/4.01M [00:00<00:00, 28.4MB/s]\n",
            "  0% 0/1 [00:00<?, ?it/s]\n",
            "  0% 0/2 [00:00<?, ?it/s]\u001b[A\n",
            " 50% 1/2 [00:01<00:01,  1.30s/it]\u001b[A\n",
            "{'load_data': '0.267', 'extract_feat': '0.911', 'forward': '1.296', 'batch_size': '1', 'rtf': '0.074'}, :  50% 1/2 [00:01<00:01,  1.30s/it]\u001b[A\n",
            "rtf_avg: 0.074: 100% 2/2 [00:01<00:00,  1.54it/s]\n",
            "time cost vad: 1.301\n",
            "\n",
            "  0% 0/2 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "  0% 0/10 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "rtf_avg_per_sample: 0.028, time_speech_total_per_sample:  77.415, time_escape_total_per_sample: 2.190:  90% 9/10 [00:02<00:00,  4.11it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "  0% 0/2 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            " 50% 1/2 [00:00<00:00,  6.85it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "{'load_data': 0.0, 'extract_feat': 0.0, 'forward': '0.146', 'batch_size': '1', 'rtf': '-0.146'}, :  50% 1/2 [00:00<00:00,  6.85it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "rtf_avg: -0.146: 100% 2/2 [00:00<00:00, 13.63it/s]\n",
            "\n",
            " 50% 1/2 [00:02<00:02,  2.44s/it]\u001b[A\n",
            "rtf_avg_all_samples: 0.032, time_speech_total_all_samples:  77.415, time_escape_total_all_samples: 2.443: 100% 2/2 [00:02<00:00,  1.22s/it]\n",
            "rtf_avg_per_sample: 0.028, time_speech_total_per_sample:  77.415, time_escape_total_per_sample: 2.190:  90% 9/10 [00:02<00:00,  3.85it/s]\n",
            "100% 1/1 [00:03<00:00,  3.75s/it]\n",
            "ASR 任务完成->标注文件路径: /content/GPT-SoVITS/output/asr_opt/slicer.list\n",
            "\n",
            "\"/usr/bin/python3\" tools/subfix_webui.py --load_list \"/content/drive/MyDrive/voices/asr_opt/slicer.list\" --webui_port 9871 --is_share True\n",
            "Running on public URL: https://b300978e3a18ceea6e.gradio.live\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/1-get-text.py\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/1-get-text.py\n",
            "[nltk_data] Downloading package averaged_perceptron_tagger to\n",
            "[nltk_data]     /root/nltk_data...\n",
            "[nltk_data] Downloading package averaged_perceptron_tagger to\n",
            "[nltk_data]     /root/nltk_data...\n",
            "[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/g2p_en/g2p.py\", line 22, in <module>\n",
            "    nltk.data.find('taggers/averaged_perceptron_tagger.zip')\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/nltk/data.py\", line 583, in find\n",
            "    raise LookupError(resource_not_found)\n",
            "LookupError: \n",
            "**********************************************************************\n",
            "  Resource \u001b[93maveraged_perceptron_tagger\u001b[0m not found.\n",
            "  Please use the NLTK Downloader to obtain the resource:\n",
            "\n",
            "  \u001b[31m>>> import nltk\n",
            "  >>> nltk.download('averaged_perceptron_tagger')\n",
            "  \u001b[0m\n",
            "  For more information see: https://www.nltk.org/data.html\n",
            "\n",
            "  Attempted to load \u001b[93mtaggers/averaged_perceptron_tagger.zip\u001b[0m\n",
            "\n",
            "  Searched in:\n",
            "    - '/root/nltk_data'\n",
            "    - '/usr/nltk_data'\n",
            "    - '/usr/share/nltk_data'\n",
            "    - '/usr/lib/nltk_data'\n",
            "    - '/usr/share/nltk_data'\n",
            "    - '/usr/local/share/nltk_data'\n",
            "    - '/usr/lib/nltk_data'\n",
            "    - '/usr/local/lib/nltk_data'\n",
            "**********************************************************************\n",
            "\n",
            "\n",
            "During handling of the above exception, another exception occurred:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/prepare_datasets/1-get-text.py\", line 18, in <module>\n",
            "    from text.cleaner import clean_text\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/text/cleaner.py\", line 1, in <module>\n",
            "    from text import chinese, japanese, cleaned_text_to_sequence, symbols, english\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/text/english.py\", line 4, in <module>\n",
            "    from g2p_en import G2p\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/g2p_en/__init__.py\", line 1, in <module>\n",
            "    from .g2p import G2p\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/g2p_en/g2p.py\", line 24, in <module>\n",
            "    nltk.download('averaged_perceptron_tagger')\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/nltk/downloader.py\", line 777, in download\n",
            "    for msg in self.incr_download(info_or_id, download_dir, force):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/nltk/downloader.py\", line 642, in incr_download\n",
            "    yield from self._download_package(info, download_dir, force)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/nltk/downloader.py\", line 735, in _download_package\n",
            "    for msg in _unzip_iter(filepath, zipdir, verbose=False):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/nltk/downloader.py\", line 2254, in _unzip_iter\n",
            "    zf.extractall(root)\n",
            "  File \"/usr/lib/python3.10/zipfile.py\", line 1647, in extractall\n",
            "    self._extract_member(zipinfo, path, pwd)\n",
            "  File \"/usr/lib/python3.10/zipfile.py\", line 1702, in _extract_member\n",
            "    shutil.copyfileobj(source, target)\n",
            "  File \"/usr/lib/python3.10/shutil.py\", line 195, in copyfileobj\n",
            "    buf = fsrc_read(length)\n",
            "  File \"/usr/lib/python3.10/zipfile.py\", line 927, in read\n",
            "    data = self._read1(n)\n",
            "  File \"/usr/lib/python3.10/zipfile.py\", line 995, in _read1\n",
            "    data += self._read2(n - len(data))\n",
            "  File \"/usr/lib/python3.10/zipfile.py\", line 1030, in _read2\n",
            "    raise EOFError\n",
            "EOFError\n",
            "[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.\n",
            "[nltk_data] Downloading package cmudict to /root/nltk_data...\n",
            "[nltk_data]   Unzipping corpora/cmudict.zip.\n",
            "Traceback (most recent call last):\n",
            "  File \"/content/GPT-SoVITS/webui.py\", line 555, in open1abc\n",
            "    with open(txt_path, \"r\",encoding=\"utf8\") as f:\n",
            "FileNotFoundError: [Errno 2] No such file or directory: 'logs/builder/2-name2text-0.txt'\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/1-get-text.py\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/1-get-text.py\n",
            "Building prefix dict from the default dictionary ...\n",
            "Dumping model to file cache /content/GPT-SoVITS/TEMP/jieba.cache\n",
            "Loading model cost 1.256 seconds.\n",
            "Prefix dict has been built succesfully.\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/2-get-hubert-wav32k.py\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/2-get-hubert-wav32k.py\n",
            "Some weights of the model checkpoint at GPT_SoVITS/pretrained_models/chinese-hubert-base were not used when initializing HubertModel: ['encoder.pos_conv_embed.conv.weight_g', 'encoder.pos_conv_embed.conv.weight_v']\n",
            "- This IS expected if you are initializing HubertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
            "- This IS NOT expected if you are initializing HubertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n",
            "Some weights of the model checkpoint at GPT_SoVITS/pretrained_models/chinese-hubert-base were not used when initializing HubertModel: ['encoder.pos_conv_embed.conv.weight_g', 'encoder.pos_conv_embed.conv.weight_v']\n",
            "- This IS expected if you are initializing HubertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
            "- This IS NOT expected if you are initializing HubertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n",
            "Some weights of HubertModel were not initialized from the model checkpoint at GPT_SoVITS/pretrained_models/chinese-hubert-base and are newly initialized: ['encoder.pos_conv_embed.conv.parametrizations.weight.original0', 'encoder.pos_conv_embed.conv.parametrizations.weight.original1']\n",
            "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
            "Some weights of HubertModel were not initialized from the model checkpoint at GPT_SoVITS/pretrained_models/chinese-hubert-base and are newly initialized: ['encoder.pos_conv_embed.conv.parametrizations.weight.original0', 'encoder.pos_conv_embed.conv.parametrizations.weight.original1']\n",
            "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n",
            "\u001b[0m\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/3-get-semantic.py\n",
            "\"/usr/bin/python3\" GPT_SoVITS/prepare_datasets/3-get-semantic.py\n",
            "/usr/local/lib/python3.10/dist-packages/torch/nn/utils/weight_norm.py:28: UserWarning: torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.\n",
            "  warnings.warn(\"torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.\")\n",
            "/usr/local/lib/python3.10/dist-packages/torch/nn/utils/weight_norm.py:28: UserWarning: torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.\n",
            "  warnings.warn(\"torch.nn.utils.weight_norm is deprecated in favor of torch.nn.utils.parametrizations.weight_norm.\")\n",
            "<All keys matched successfully>\n",
            "<All keys matched successfully>\n",
            "DEBUG:jax._src.path:etils.epath found. Using etils.epath for file I/O.\n",
            "\"/usr/bin/python3\" GPT_SoVITS/s2_train.py --config \"/content/GPT-SoVITS/TEMP/tmp_s2.json\"\n",
            "2024-04-05 05:17:23.672440: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "2024-04-05 05:17:23.672492: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "2024-04-05 05:17:23.673778: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "DEBUG:tensorflow:Falling back to TensorFlow client; we recommended you install the Cloud TPU client directly with pip install cloud-tpu-client.\n",
            "2024-04-05 05:17:25.349108: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
            "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
            "DEBUG:h5py._conv:Creating converter from 5 to 7\n",
            "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
            "DEBUG:h5py._conv:Creating converter from 5 to 7\n",
            "DEBUG:jax._src.path:etils.epath found. Using etils.epath for file I/O.\n",
            "INFO:numexpr.utils:NumExpr defaulting to 2 threads.\n",
            "2024-04-05 05:17:30.534684: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "2024-04-05 05:17:30.534739: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "2024-04-05 05:17:30.536185: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "DEBUG:tensorflow:Falling back to TensorFlow client; we recommended you install the Cloud TPU client directly with pip install cloud-tpu-client.\n",
            "2024-04-05 05:17:31.667313: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
            "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
            "DEBUG:h5py._conv:Creating converter from 5 to 7\n",
            "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
            "DEBUG:h5py._conv:Creating converter from 5 to 7\n",
            "DEBUG:jax._src.path:etils.epath found. Using etils.epath for file I/O.\n",
            "INFO:numexpr.utils:NumExpr defaulting to 2 threads.\n",
            "INFO:dean:{'train': {'log_interval': 100, 'eval_interval': 500, 'seed': 1234, 'epochs': 8, 'learning_rate': 0.0001, 'betas': [0.8, 0.99], 'eps': 1e-09, 'batch_size': 7, 'fp16_run': True, 'lr_decay': 0.999875, 'segment_size': 20480, 'init_lr_ratio': 1, 'warmup_epochs': 0, 'c_mel': 45, 'c_kl': 1.0, 'text_low_lr_rate': 0.4, 'pretrained_s2G': 'GPT_SoVITS/pretrained_models/s2G488k.pth', 'pretrained_s2D': 'GPT_SoVITS/pretrained_models/s2D488k.pth', 'if_save_latest': True, 'if_save_every_weights': True, 'save_every_epoch': 4, 'gpu_numbers': '0'}, 'data': {'max_wav_value': 32768.0, 'sampling_rate': 32000, 'filter_length': 2048, 'hop_length': 640, 'win_length': 2048, 'n_mel_channels': 128, 'mel_fmin': 0.0, 'mel_fmax': None, 'add_blank': True, 'n_speakers': 300, 'cleaned_text': True, 'exp_dir': 'logs/dean'}, 'model': {'inter_channels': 192, 'hidden_channels': 192, 'filter_channels': 768, 'n_heads': 2, 'n_layers': 6, 'kernel_size': 3, 'p_dropout': 0.1, 'resblock': '1', 'resblock_kernel_sizes': [3, 7, 11], 'resblock_dilation_sizes': [[1, 3, 5], [1, 3, 5], [1, 3, 5]], 'upsample_rates': [10, 8, 2, 2, 2], 'upsample_initial_channel': 512, 'upsample_kernel_sizes': [16, 16, 8, 2, 2], 'n_layers_q': 3, 'use_spectral_norm': False, 'gin_channels': 512, 'semantic_frame_rate': '25hz', 'freeze_quantizer': True}, 's2_ckpt_dir': 'logs/dean', 'content_module': 'cnhubert', 'save_weight_dir': 'SoVITS_weights', 'name': 'dean', 'pretrain': None, 'resume_step': None}\n",
            "phoneme_data_len: 1\n",
            "wav_data_len: 100\n",
            "100% 100/100 [00:00<00:00, 16873.05it/s]\n",
            "skipped_phone:  0 , skipped_dur:  100\n",
            "total left:  0\n",
            "Traceback (most recent call last):\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/s2_train.py\", line 600, in <module>\n",
            "    main()\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/s2_train.py\", line 56, in main\n",
            "    mp.spawn(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py\", line 241, in spawn\n",
            "    return start_processes(fn, args, nprocs, join, daemon, start_method=\"spawn\")\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py\", line 197, in start_processes\n",
            "    while not context.join():\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py\", line 158, in join\n",
            "    raise ProcessRaisedException(msg, error_index, failed_process.pid)\n",
            "torch.multiprocessing.spawn.ProcessRaisedException: \n",
            "\n",
            "-- Process 0 terminated with the following error:\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/torch/multiprocessing/spawn.py\", line 68, in _wrap\n",
            "    fn(i, *args)\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/s2_train.py\", line 85, in run\n",
            "    train_dataset = TextAudioSpeakerLoader(hps.data)  ########\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/module/data_utils.py\", line 103, in __init__\n",
            "    assert len(audiopaths_sid_text_new) > 1  # 至少能凑够batch size，这里todo\n",
            "AssertionError\n",
            "\n",
            "\"/usr/bin/python3\" GPT_SoVITS/s1_train.py --config_file \"/content/GPT-SoVITS/TEMP/tmp_s1.yaml\" \n",
            "Seed set to 1234\n",
            "Using 16bit Automatic Mixed Precision (AMP)\n",
            "GPU available: True (cuda), used: True\n",
            "TPU available: False, using: 0 TPU cores\n",
            "IPU available: False, using: 0 IPUs\n",
            "HPU available: False, using: 0 HPUs\n",
            "<All keys matched successfully>\n",
            "ckpt_path: None\n",
            "[rank: 0] Seed set to 1234\n",
            "Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1\n",
            "----------------------------------------------------------------------------------------------------\n",
            "distributed_backend=nccl\n",
            "All distributed processes registered. Starting with 1 processes\n",
            "----------------------------------------------------------------------------------------------------\n",
            "\n",
            "Missing logger folder: logs/dean/logs_s1/logs_s1\n",
            "2024-04-05 05:17:58.841531: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "2024-04-05 05:17:58.841594: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "2024-04-05 05:17:58.842886: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2024-04-05 05:17:59.956005: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
            "semantic_data_len: 1\n",
            "phoneme_data_len: 1\n",
            "                             item_name                                     semantic_audio\n",
            "0  title.wav_0000000000_0002477440.wav  536 99 965 273 750 581 103 318 318 451 722 726...\n",
            "Traceback (most recent call last):\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/s1_train.py\", line 171, in <module>\n",
            "    main(args)\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/s1_train.py\", line 147, in main\n",
            "    trainer.fit(model, data_module, ckpt_path=ckpt_path)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py\", line 544, in fit\n",
            "    call._call_and_handle_interrupt(\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py\", line 43, in _call_and_handle_interrupt\n",
            "    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/strategies/launchers/subprocess_script.py\", line 105, in launch\n",
            "    return function(*args, **kwargs)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py\", line 580, in _fit_impl\n",
            "    self._run(model, ckpt_path=ckpt_path)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/trainer.py\", line 949, in _run\n",
            "    call._call_setup_hook(self)  # allow user to set up LightningModule in accelerator environment\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py\", line 92, in _call_setup_hook\n",
            "    _call_lightning_datamodule_hook(trainer, \"setup\", stage=fn)\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/call.py\", line 179, in _call_lightning_datamodule_hook\n",
            "    return fn(*args, **kwargs)\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/AR/data/data_module.py\", line 29, in setup\n",
            "    self._train_dataset = Text2SemanticDataset(\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/AR/data/dataset.py\", line 107, in __init__\n",
            "    self.init_batch()\n",
            "  File \"/content/GPT-SoVITS/GPT_SoVITS/AR/data/dataset.py\", line 187, in init_batch\n",
            "    for _ in range(max(2, int(min_num / leng))):\n",
            "ZeroDivisionError: division by zero\n"
          ]
        }
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}